### PR TITLE
TEP-0100: Embedded TaskRuns and Runs Status in PipelineRuns [Problem Statement]

### PR DESCRIPTION
**Pull request by jerop**
This TEP proposes changes to `PipelineRun` Status to reduce the amount of
information stored about the status of its `Run`s and `TaskRun`s to improve
performance, reduce memory bloat and improve extensibility.

Related Issues:
- [PipelineRun and TaskRun Status][issue-3792]
- [PipelineRun Status bloat and excessive updates][issue-3140]

Related TEPs:
- [Pipelines in Pipelines][tep-0056]
- [Matrix][tep-0090]

/kind tep

Co-authored-by: Lee Bernick <leebernick@google.com>

[tep-0056]: https://github.com/tektoncd/community/blob/main/teps/0056-pipelines-in-pipelines.md
[tep-0090]: https://github.com/tektoncd/community/blob/main/teps/0090-matrix.md
[issue-3140]: https://github.com/tektoncd/pipeline/issues/3140
[issue-3792]: https://github.com/tektoncd/pipeline/issues/3792

### CODE REVIEW

**Comment by jerop** [comment]
/cc @skaegi @imjasonh @bobcatfish @sbwsg @abayer @vdemeester @lbernick 

(from previous discussions about pipelinerun status in https://github.com/tektoncd/pipeline/issues/3140 and https://github.com/tektoncd/pipeline/issues/3792)

**Comment by bobcatfish** [comment]
ONE HUNDREDTH TEP!!! 🎉 🎉 🎉 

image

dj cat celebration

**Comment by jerop** [comment]
putting a hold on this PR so it doesn't get merged on the first approval to give people more time to review 

/hold

File: teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md
Comment (pritidesai) [inline teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md:66]: pipelineRun status contains `startTime`, `completionTime`, `conditions` (indicating success or failure of pipelineRun itself), `pipelineResults`, `pipelineSpec`,  a map of `taskRuns`, and a map of `runs`.

Here, we are trying to optimize the map of `taskRun` and `run`. Each `taskRun` and `run` in turn contains `startTime`, `completionTime`, `conditions` (indicating success or failure of taskRun/run itself), `podName`, a list of steps, `taskResults`, and `taskSpec`.

File: teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md
Comment (vdemeester) [inline teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md:60]: One of *today's* requirement is that it has to contain the related "resolved" definition (`Pipeline` for `PipelineRun` and `Task` for `TaskRun`) so that we don't fetch the definition each and every time. We may not want to keep that requirement but we should keep the behavior (fetch the definition only once)

**Review by vdemeester — COMMENTED** [review]
/assign

File: teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md
Comment (chmouel) [reply teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md:66]: it may be nice to mention what are we "loosing" out of the optimizations, which fields are going to be not available anymore for tools/users to parse.

an example of the expected status would be nice too,

**Comment by bobcatfish** [comment]
no particular feedback from me!

/approve

File: teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md
Comment (jerop) [reply teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md]: Thank you for the reviews @pritidesai and @chmouel!

@pritidesai added a background section providing more information about the `PipelineRunStatus` as you described above

@chmouel we plan to discuss what fields will be removed and provide examples in the proposal (which will be added in the next pull request) - added an "open questions" section with your question above to ensure they will be addressed in the proposal

File: teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md
Comment (jerop) [reply teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md]: The resolved `PipelineSpec` is out of scope for this TEP (only targeting the `TaskRuns` and `Runs`) - added a background section clarifying this :)

**Comment by jerop** [comment]
@imjasonh @chmouel @pritidesai @vdemeester thanks for the initial reviews - clarified the problem statement to address your comments, please take a look 🙏🏾

**Review by vdemeester — APPROVED** [review]

**Comment by chmouel** [comment]
LGTM

**Pull request merged**
